File: app/src/main/java/com/android/example/github/MainViewModel.java
package com.android.example.github;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModel;

import com.android.example.github.repository.StoryRepository;
import com.android.example.github.repository.UserRepository;
import com.android.example.github.vo.Resource;
import com.android.example.github.vo.User;

import java.io.File;

import javax.inject.Inject;

import kotlin.Pair;

public class MainViewModel extends ViewModel {

    private final UserRepository userRepository;
    private final StoryRepository storyRepository;

    @Inject
    MainViewModel(StoryRepository repository, UserRepository userRepository) {
        this.userRepository = userRepository;
        this.storyRepository = repository;
    }

    LiveData<Resource<User>> getUser(String userId) {
        return userRepository.loadUser(userId);
    }

    LiveData<Resource<Void>> createUser(User user) {
        return userRepository.putUser(user);
    }

    LiveData<Resource<String>> putFileInS3(Pair<String, File> stringFilePair) {
        return storyRepository.putFileInS3(stringFilePair);
    }
}

File: app/src/main/java/com/talkingwhale/MainViewModel.kt
package com.talkingwhale

import android.arch.core.util.Function
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations
import android.arch.lifecycle.ViewModel
import android.content.Context
import com.talkingwhale.repository.PostRepository
import com.talkingwhale.repository.UserRepository
import com.talkingwhale.vo.Post
import com.talkingwhale.vo.Resource
import com.talkingwhale.vo.User


class MainViewModel : ViewModel() {

    private val postRepository = PostRepository
    private val userRepository = UserRepository
    var currentUser: User? = null
    var localPosts: LiveData<Resource<List<Post>>> = MutableLiveData()
    private var getPosts: MutableLiveData<PostRepository.CornerLatLng> = MutableLiveData()

    init {
        localPosts = Transformations.switchMap(getPosts, Function {
            return@Function postRepository.getNearbyPosts(it)
        })
    }

    fun getUser(userId: String): LiveData<Resource<User>> {
        return userRepository.loadUser(userId)
    }

    fun putUser(user: User): LiveData<Resource<Unit>> {
        return userRepository.putUser(user)
    }

    fun putPost(post: Post): LiveData<Resource<Unit>> {
        return postRepository.addPost(post)
    }

    fun putFile(pair: Pair<Post, Context>): LiveData<Resource<Post>> {
        return postRepository.putFile(pair)
    }

    fun getNewPosts(cornerLatLng: PostRepository.CornerLatLng) {
        getPosts.value = cornerLatLng
    }

    fun deletePost(post: Post): LiveData<Resource<Unit>> {
        return postRepository.deletePost(post)
    }
}